web/src/narrow_state.ts

```diff
@@ -265,6 +268,36 @@ export function _possible_unread_message_ids(
     let topic_name;
     let current_filter_pm_string;
 
+    // For the `with` operator, we can only correctly compute the
+    // correct channel/topic for lookup unreads in if we either
+    // have the message in our local cache, or we know the filter
+    // has already been updated for potentially moved messages.
+    //
+    // The code path that needs this function is never called in
+    // the `with` code path, but for safety, we assert that
+    // assumption is not violated.
+    //
+    // If we need to change that assumption, we can try looking up the
+    // target message in message_store, but would need to return
+    // undefined if the target message is not available.
+    assert(!current_filter.requires_adjustment_for_moved_with_target);
+
+    if (current_filter.can_bucket_by("channel", "topic", "with")) {
+        sub = stream_sub(current_filter);
+        topic_name = topic(current_filter);
+
+        const with_operand = current_filter.operands("with")[0]!;
+        const target_id = Number.parseInt(with_operand, 10);
+        const target_message = message_store.get(target_id)!;
+
+        // Since direct messages can not be moved to channels
+        // it would be a noop
+        if (target_message.type !== "stream" || sub === undefined || topic_name === undefined) {
+            return [];
+        }
```

Comment:
direct messages containing channel or topic narrow will also be leaked into this function since `requires_adjustment_for_moved_with_target` is `false`. Hence, I added them as a noop. I'll update the comment if it doesn't look good.

Reply:
This has two potential bugs:
- First, `target_message` might be undefined if the message ID is not in the client cache; this code will crash in that case.
- If `target_message` is a DM conversation, I don't think there's a situation where we'd want to just return an empty list; we should be trying to return the unreads in the target conversation. I think that might be an unexpected situation where the client may be unable to calculate things correctly and returning `undefined` might be appropriate?

Reply:
So I think the actual bug here is that `adjusted_terms_if_moved` doesn't do the correct thing for DM messages. I added this comment explaining it:

```
    static adjusted_terms_if_moved(raw_terms: NarrowTerm[], message: Message): NarrowTerm[] | null {                                  
        if (message.type !== "stream") {                                                                                              
            // BUG: We should be replacing the channel/topic/dm                                                                       
            // operators with the singular dm operator for the                                                                        
            // conversation in question. This isn't very important,                                                                   
            // because direct messages cannot be moved, but it should                                                                 
            // be fixed, since it is triggerable incorrect behavior.                                                                  
            return null;                                                                                                              
        }                                                                                                                             
```

This should instead be matching the algorithm that we have in the backend, in `update_narrow_terms_containing_with_operator`, which would result in a guarantee that the operators post-processing (when `!requires_adjustment_for_moved_with_target`) will never have a direct message paired with channel/topic operators.

Reply:
This looks like a satisfactory temporary change for this, but please put up a PR fixing the core bug here.

``` diff
diff --git a/web/src/filter.ts b/web/src/filter.ts
index 0da766cbee..e226dbeb55 100644
--- a/web/src/filter.ts
+++ b/web/src/filter.ts
@@ -808,6 +808,11 @@ export class Filter {
 
     static adjusted_terms_if_moved(raw_terms: NarrowTerm[], message: Message): NarrowTerm[] | null {
         if (message.type !== "stream") {
+            // BUG: We should be replacing the channel/topic/dm
+            // operators with the singular dm operator for the
+            // conversation in question. This isn't very important,
+            // because direct messages cannot be moved, but it should
+            // be fixed, since it is triggerable incorrect behavior.
             return null;
         }
 
diff --git a/web/src/narrow_state.ts b/web/src/narrow_state.ts
index fdb57276b1..f2acbfd77b 100644
--- a/web/src/narrow_state.ts
+++ b/web/src/narrow_state.ts
@@ -290,9 +290,15 @@ export function _possible_unread_message_ids(
         const target_id = Number.parseInt(with_operand, 10);
         const target_message = message_store.get(target_id)!;
 
-        // Since direct messages can not be moved to channels
-        // it would be a noop
-        if (target_message.type !== "stream" || sub === undefined || topic_name === undefined) {
+        if (target_message?.type === "direct") {
+            // BUG: In theory, the fact that we've asserted
+            // !current_filter.requires_adjustment_for_moved_with_target
+            // should mean this is not possible; but
+            // filter.adjusted_terms_if_moved incorrectly does not
+            // ensure this. Once that bug is fixed, we can delete this case.
+            return [];
+        }
+        if (sub === undefined || topic_name === undefined) {
             return [];
         }
         return unread.get_msg_ids_for_topic(sub.stream_id, topic_name);
```

Reply:
I'll update it in some time